sw/device/tests/kmac_idle_test.c

```diff
@@ -0,0 +1,131 @@
+  CHECK(clock_state == expected_clock_state,
+        "Clock enabled state is not as expected (%0d).", expected_clock_state);
+}
+
+bool test_main(void) {
+  // clkmgr init.
+  CHECK_DIF_OK(dif_clkmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_CLKMGR_AON_BASE_ADDR), &clkmgr));
+
+  // Get initial hint and enable for KMAC clock and check both are enabled.
+  dif_toggle_t clock_hint_state;
+  dif_toggle_t clock_state;
+
+  CHECK_DIF_OK(
+      dif_clkmgr_hintable_clock_get_hint(&clkmgr, clock, &clock_hint_state));
+  CHECK_DIF_OK(
+      dif_clkmgr_hintable_clock_get_enabled(&clkmgr, clock, &clock_state));
+  CHECK((clock_state == kDifToggleEnabled) &&
+            (clock_hint_state == kDifToggleEnabled),
+        "Initial clock hint and/or clock state for KMAC not enabled.");
+
+  // While KMAC is not in use, set disabled hint for KMAC clock
+  // then check clock state, set back to enabled and check again.
+  // Note: disabled means the clock can be disabled (i.e the hint is set).
+  set_hint_check_clock_state(kDifToggleDisabled, kDifToggleDisabled);
+  set_hint_check_clock_state(kDifToggleEnabled, kDifToggleEnabled);
+
+  // Initialize KMAC hardware.
+  dif_kmac_operation_state_t kmac_operation_state;
+  CHECK_DIF_OK(
+      dif_kmac_init(mmio_region_from_addr(TOP_EARLGREY_KMAC_BASE_ADDR), &kmac));
+
+  // Configure KMAC hardware using software entropy.
+  dif_kmac_config_t config = (dif_kmac_config_t){
+      .entropy_mode = kDifKmacEntropyModeSoftware,
+      .entropy_seed = 0xffff,
+      .entropy_fast_process = kDifToggleEnabled,
+  };
+  CHECK_DIF_OK(dif_kmac_configure(&kmac, config));
+
+  // Run SHA3 test case using single blocking absorb/squeeze operations.
+  CHECK_DIF_OK(dif_kmac_mode_sha3_start(&kmac, &kmac_operation_state,
+                                        sha3_256_test.mode));
+
+  // Set hint and check clk state remains enabled as KMAC is now not idle.
+  set_hint_check_clock_state(kDifToggleDisabled, kDifToggleEnabled);
+
+  CHECK_DIF_OK(dif_kmac_absorb(&kmac, &kmac_operation_state,
+                               sha3_256_test.message, sha3_256_test.message_len,
+                               NULL));
+
+  uint32_t out[DIGEST_LEN_SHA3_MAX];
+  CHECK_DIF_OK(dif_kmac_squeeze(&kmac, &kmac_operation_state, out,
+                                sha3_256_test.digest_len, NULL));
+
+  // Set hint and check clock state again which should still be enabled.
+  set_hint_check_clock_state(kDifToggleDisabled, kDifToggleEnabled);
+
+  // Before ending the operation set hint to enabled state,
+  // this is necessary because the end function sets the done command,
+  // at this point the KMAC will go idle and the clock will stop.
+  // However, the second part of the end function polls the state register
+  // checking for the idle state which would lock up without a clock.
+  set_hint_check_clock_state(kDifToggleEnabled, kDifToggleEnabled);
+
+  CHECK_DIF_OK(dif_kmac_end(&kmac, &kmac_operation_state));
```

Comment:
Maybe you could update the [`dif_kmac_end`](https://github.com/lowRISC/opentitan/blob/72201cf1076d97016344242b3dfa8f965b7be8a6/sw/device/lib/dif/dif_kmac.c#L732) DIF to be more similar to the [`dif_hmac_finish` ](https://github.com/lowRISC/opentitan/blob/72201cf1076d97016344242b3dfa8f965b7be8a6/sw/device/lib/dif/dif_hmac.c#L244) DIF and be non-blocking (i.e., just remove [lines 762--768](https://github.com/lowRISC/opentitan/blob/72201cf1076d97016344242b3dfa8f965b7be8a6/sw/device/lib/dif/dif_kmac.c#L762-L768)), so this work-around is not needed?